src/v/kafka/server/handlers/handler_probe.cc

```diff
@@ -0,0 +1,123 @@
+    for (size_t i = 0; i < _probes.size(); i++) {
+        auto key = api_key{i};
+
+        if (handler_for_key(key) || i == _unknown_handler_index) {
+            _probes[i] = std::make_unique<handler_probe>(key);
+            _probes.at(i)->setup_metrics();
+        }
+    }
+}
+
+handler_probe& handler_probe_manager::get_probe(api_key key) {
+    if (!handler_for_key(key)) {
+        key = api_key{_unknown_handler_index};
+    }
+
+    return *_probes.at(key);
+}
+
+handler_probe::handler_probe(api_key key)
+  : _key(key) {}
+
+void handler_probe::setup_metrics() {
+    namespace sm = ss::metrics;
+
+    if (config::shard_local_cfg().disable_metrics()) {
+        return;
+    }
+
+    const char* handler_name;
+    if (auto handler = handler_for_key(_key)) {
+        handler_name = handler.value()->name();
+    } else {
+        handler_name = "unknown_handler";
+    }
+
+    std::vector<sm::label_instance> labels{sm::label("handler")(handler_name)};
+    auto aggregate_labels = config::shard_local_cfg().aggregate_metrics()
+                              ? std::vector<sm::label>{sm::shard_label}
+                              : std::vector<sm::label>{};
+
+    _metrics.add_group(
+      prometheus_sanitize::metrics_name("kafka:handler"),
+      {
+        sm::make_counter(
+          "requests_completed",
+          [this] { return _requests_completed; },
+          sm::description("Number of requests completed"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "requests_errored",
+          [this] { return _requests_errored; },
+          sm::description("Number of requests errored"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "requests_in_progress",
+          [this] { return _requests_in_progress_every_ns / 1'000'000'000; },
+          sm::description("Number of requests in progress"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "bytes_received",
+          [this] { return _bytes_received; },
+          sm::description("Number of bytes received from requests"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "bytes_sent",
+          [this] { return _bytes_sent; },
+          sm::description("Number of bytes sent in replies"),
+          labels)
+          .aggregate(aggregate_labels),
+      });
+}
+
+/*
+ * This roughly approximates an integral of `_requests_in_progress`.
+ * By providing Prometheus with a counter of the integral rather than
+ * a gauge for `_requests_in_progress` we avoid any bias in the value
+ * that Prometheus's sampling rate could introduce.
+ */
+void handler_probe::sample_in_progress() {
+    auto now = ss::lowres_clock::now();
+    auto s_diff = std::chrono::duration_cast<std::chrono::nanoseconds>(
+                    now - _last_recorded_in_progress)
+                    .count();
+
+    if (s_diff > 0) {
```

Comment:
Yeah, before I was updating the values with much less frequency(1ms) so the conditional was needed to span the gap between samples. Now that the frequency is 1ns I suppose this isn't needed anymore. Will remove.

Reply:
Regardless of the frequency, the code inside the `if` does nothing if `s_diff == 0`, right? 

```
_requests_in_progress_every_ns += _requests_in_progress * s_diff;
```

ends up as `+= 0` since s_diff is zero

```
_last_recorded_in_progress = now;
```

is a no op since `_last_recorded_in_progress == now` when `s_diff` is 0.

So there is no special case which makes sense since it's not obvious why a zero-size time slice should be special.